pkg/events/derive/so_export_symbol.go

```diff
@@ -0,0 +1,141 @@
+	"github.com/aquasecurity/tracee/types/trace"
+	"path"
+	"strings"
+)
+
+func SharedObjectExportWatchedSymbols(soLoader sharedobjs.DynamicSymbolsLoader, watchedSymbols []string, whitelistedLibsPrefixes []string) events.DeriveFunction {
+	gen := initSOExportWatchedSymbolsEventGenerator(soLoader, watchedSymbols, whitelistedLibsPrefixes)
+	return singleEventDeriveFunc(events.SymbolsLoad, gen.GenerateDeriveArgsFunc)
+}
+
+// Most specific paths should be at the top, to prevent bugs with iterations over the list
+var knownLibrariesDirs = []string{
+	"/usr/lib/x86_64-linux-gnu/",
+	"/usr/lib64/",
+	"/usr/lib/",
+	"/lib64/",
+	"/lib/",
+}
+
+// soExportWatchedSymbolsEventGenerator is responsible of generating event if shared object loaded to a process
+// export one or more from given watched sybmols.
+type soExportWatchedSymbolsEventGenerator struct {
+	soLoader            sharedobjs.DynamicSymbolsLoader
+	watchedSymbols      map[string]bool
+	pathPrefixWhitelist []string
+	librariesWhitelist  []string
+}
+
+func initSOExportWatchedSymbolsEventGenerator(
+	soLoader sharedobjs.DynamicSymbolsLoader,
+	watchedSymbols []string,
+	whitelistedLibsPrefixes []string) *soExportWatchedSymbolsEventGenerator {
+	watchedSymbolsMap := make(map[string]bool)
+	for _, sym := range watchedSymbols {
+		watchedSymbolsMap[sym] = true
+	}
+	var libraries, prefixes []string
+	for _, path := range whitelistedLibsPrefixes {
+		if strings.HasPrefix(path, "/") {
+			prefixes = append(prefixes, path)
+		} else {
+			libraries = append(libraries, path)
+		}
+	}
+	return &soExportWatchedSymbolsEventGenerator{
+		soLoader:            soLoader,
+		watchedSymbols:      watchedSymbolsMap,
+		pathPrefixWhitelist: prefixes,
+		librariesWhitelist:  libraries,
+	}
+}
+
+func (soExSymbolGen *soExportWatchedSymbolsEventGenerator) GenerateDeriveArgsFunc(event trace.Event) ([]interface{}, error) {
+	loadingObjectInfo, err := getSharedObjectExInfo(event)
+	if err != nil {
+		return nil, err
+	}
+
+	if soExSymbolGen.isWhitelist(loadingObjectInfo.Path) {
+		return nil, nil
+	}
+
+	soSyms, err := soExSymbolGen.soLoader.GetExportedSymbols(loadingObjectInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	var exportedWatchSymbols []string
+	for sym := range soSyms {
+		if soExSymbolGen.watchedSymbols[sym] {
+			exportedWatchSymbols = append(exportedWatchSymbols, sym)
+		}
+	}
+
+	if len(exportedWatchSymbols) > 0 {
+		return []interface{}{loadingObjectInfo.Path, exportedWatchSymbols}, nil
+	} else {
+		return nil, nil
+	}
+}
+
+// isWhitelist check if a SO's path is in the whitelist given in initialization
+func (soExSymbolGen *soExportWatchedSymbolsEventGenerator) isWhitelist(soPath string) bool {
+	// Check absolute path libraries whitelist
+	for _, prefix := range soExSymbolGen.pathPrefixWhitelist {
+		if strings.HasPrefix(soPath, prefix) {
+			return true
+		}
+	}
+
+	// Check if SO is whitelisted library which resides in one of the known libs paths
+	if len(soExSymbolGen.librariesWhitelist) > 0 {
```

Comment:
However, the `knownLibrariesDirs` slices is in a fixed short size.
On the other hand, there might be many libraries given.
To reduce performance, this way is most efficient because we only check libraries if the path is relevant.